lib/pbench/test/unit/server/test_cache_manager.py

```diff
@@ -1863,3 +1890,58 @@ def fake_open(self, flags: str) -> FakeStream:
         assert lock.lock is None
         assert files == [("open", lockfile, "w+")]
         assert locks == [(lockfile, fcntl.LOCK_SH), (lockfile, fcntl.LOCK_UN)]
+
+    def test_lockmanager(self, monkeypatch):
+        """Test LockManager"""
+
+        lockfile = Path("/lock")
+        monkeypatch.setattr("pbench.server.cache_manager.LockRef", FakeLockRef)
+        # default parameters
+        with LockManager(lockfile) as lock:
+            assert lock.unlock
+            assert not lock.exclusive
+            assert lock.wait
+            assert FakeLockRef.operations == [("acquire", False, True)]
+            lock.upgrade()
+            assert FakeLockRef.operations == [("acquire", False, True), ("upgrade")]
+            lock.downgrade()
+            assert FakeLockRef.operations == [
+                ("acquire", False, True),
+                ("upgrade"),
+                ("downgrade"),
+            ]
+        assert FakeLockRef.operations == [
+            ("acquire", False, True),
+            ("upgrade"),
+            ("downgrade"),
+            ("release"),
+        ]
+
+        # exclusive
+        FakeLockRef.reset()
+        with LockManager(lockfile, exclusive=True) as lock:
+            assert lock.unlock
+            assert lock.exclusive
+            assert lock.wait
+            assert FakeLockRef.operations == [("acquire", True, True)]
+        assert FakeLockRef.operations == [("acquire", True, True), ("release")]
+
+        # no-wait
+        FakeLockRef.reset()
+        with LockManager(lockfile, wait=False) as lock:
+            assert lock.unlock
+            assert not lock.exclusive
+            assert not lock.wait
+            assert FakeLockRef.operations == [("acquire", False, False)]
+        assert FakeLockRef.operations == [("acquire", False, False), ("release")]
+
+        # keep' option
+        FakeLockRef.reset()
+        with LockManager(lockfile) as lock:
+            assert lock.keep() is lock
+            still_locked = lock
+            assert not lock.unlock
```

Comment:
Perhaps it's a paradigm trap that I'm in, but I feel a certain revulsion at the idea that calls to `assert` might have side-effects or otherwise change state.  (Assertions should _check_ things, not change them....)  So, "hiding" a changer in an assertion is evil in that it is likely to be overlooked and sow confusion.  (Or, at least, I had to work extra hard to figure out how this block of code was (successfully) asserting `not lock.unlock` when the previous, very similar code was asserting `lock.unlock`....)  And, yes, this is test code, but don't get me started on the use of `assert` in production code....  (I had no intention of insulting line 1941...it might even be proud of being evil -- I don't know!  😈)

Reply:
`assert` is just a way to execute an expression and test the result. I might agree, philosophically, if I were testing *side effects* of the method ... but I was testing the *result* of the method and there's absolutely nothing unusual about that.